lib/JsonClient.js

```diff
@@ -57,18 +60,22 @@ JsonClient.prototype.parse = function parse(req, callback) {
     function parseResponse(err, req2, res, data) {
         var obj;
         var resErr = err;
+        var parseErr;
 
         try {
             if (data && !/^\s*$/.test(data)) {
                 obj = JSON.parse(data);
             }
         } catch (e) {
-            // Not really sure what else we can do here, besides
-            // make the client just keep going.
-            log.trace(e, 'Invalid JSON in response');
+            // bad data being returned that cannot be parsed should be surfaced
+            // to the caller. http errors should take precedence, but it's
+            // possible to receive malformed data regardless of a status code.
+            parseErr = e;
+            log.trace(parseErr, 'Invalid JSON in response');
         }
         obj = obj || {};
```

Comment:
This may be explained somewhere, but why do we return an empty object on error? Should this be `undefined`?

Reply:
Good call - this should just be `data` to return the original value where possible in the event of a parse error. That way you get both the error as well as the original (unparsed) payload.